ovh/resource_dedicated_nasha_partition_access.go

```diff
@@ -0,0 +1,143 @@
+package ovh
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"log"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceDedicatedNASHAPartitionAccess() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceDedicatedNASHAPartitionAccessCreate,
+		ReadContext:   resourceDedicatedNASHAPartitionAccessRead,
+		DeleteContext: resourceDedicatedNASHAPartitionAccessDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"service_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"partition_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"ip": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"type": &schema.Schema{
+				Type:     schema.TypeString,
+				Default:  "readwrite",
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceDedicatedNASHAPartitionAccessCreate(c context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*Config)
+	serviceName := d.Get("service_name").(string)
+	partitionName := d.Get("partition_name").(string)
+	ipsubnet, _ := d.Get("ip").(string)
+
+	access := &DedicatedNASHAPartitionAccess{
+		IP:   d.Get("ip").(string),
+		Type: d.Get("type").(string),
+	}
+
+	resp := &DedicatedNASHATask{}
+
+	endpoint := fmt.Sprintf("/dedicated/nasha/%s/partition/%s/access", serviceName, partitionName)
+
+	err := config.OVHClient.Post(endpoint, access, resp)
+	if err != nil {
+		return diag.Errorf("calling %s with params %v:\n\t %q", endpoint, access, err)
+	}
+
+	stateConf := resp.StateChangeConf(d, meta)
+
+	_, err = stateConf.WaitForStateContext(c)
+	if err != nil {
+		return diag.Errorf("waiting for HA-NAS partition access (%v): %s", access, err)
+	}
+	log.Printf("[DEBUG] Created HA-NAS partition access")
+
+	d.SetId(fmt.Sprintf("%s/%s/%s", serviceName, partitionName, url.PathEscape(ipsubnet)))
+
+	return nil
+}
+
+func resourceDedicatedNASHAPartitionAccessRead(c context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*Config)
+
+	if strings.Contains(d.Id(), "/") {
+		tab := strings.Split(d.Id(), "/")
+		if len(tab) != 3 {
+			return diag.Errorf("can't parse access partition ID: %s", d.Id())
+		}
+
+		d.Set("service_name", tab[0])
+		d.Set("partition_name", tab[1])
+		ip, _ := url.PathUnescape(tab[2])
+		d.Set("ip", ip)
+	}
+
+	serviceName := d.Get("service_name").(string)
+	partitionName := d.Get("partition_name").(string)
+	ipsubnet, _ := d.Get("ip").(string)
```

Comment:
The IP stored/read is not correct.

Indeed, when I apply an access resource:

```
ovh_dedicated_nasha_partition_access.my-partition: Creating...
ovh_dedicated_nasha_partition_access.my-partition: Still creating... [10s elapsed]
ovh_dedicated_nasha_partition_access.my-partition: Still creating... [20s elapsed]
ovh_dedicated_nasha_partition_access.my-partition: Still creating... [30s elapsed]
ovh_dedicated_nasha_partition_access.my-partition: Still creating... [40s elapsed]
ovh_dedicated_nasha_partition_access.my-partition: Still creating... [50s elapsed]
ovh_dedicated_nasha_partition_access.my-partition: Still creating... [1m0s elapsed]
ovh_dedicated_nasha_partition_access.my-partition: Creation complete after 1m1s [id=zpool-xxxx/my-partition/123.123.123.123/32]

Apply complete! Resources: 2 added, 0 changed, 1 destroyed.
```

And when I do an apply again (after solved a little the read function):

```
 $ terraform apply
data.ovh_dedicated_nasha.foo: Reading...
ovh_dedicated_nasha_partition.my-partition: Refreshing state... [id=zpool-xxxxx/my-partition]
ovh_dedicated_nasha_partition_access.my-partition: Refreshing state... [id=zpool-xxxxxxx/my-partition/123.123.123.123/32]
ovh_dedicated_nasha_partition_snapshot.my-partition: Refreshing state... [id=zpool-xxxxxx/my-partition/day-3]
data.ovh_dedicated_nasha.foo: Read complete after 1s [id=zpool-xxxxx]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # ovh_dedicated_nasha_partition_access.my-partition must be replaced
-/+ resource "ovh_dedicated_nasha_partition_access" "my-partition" {
      ~ id             = "zpool-xxxx/my-partition/123.123.123.123/32" -> (known after apply)
      ~ ip             = "123.123.123.123" -> "123.123.123.123/32" # forces replacement
        # (3 unchanged attributes hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: ^C

Interrupt received.
Please wait for Terraform to exit or data loss may occur.
Gracefully shutting down...

╷
│ Error: error asking for approval: interrupted
```